### PR TITLE
Issue355 new resets

### PR DESCRIPTION
**Pull request by kerimoyle**
Updating all parts of the code reflect the changes to the `reset` format and the removal of `when` elements (issue #355).

- [x] reset description
- [x] parser
- [x] printer
- [x] validator
- [x] coverage

NB: Reset order uniqueness (as discussed in #326) will be done in a separate PR

### CODE REVIEW

**Comment by kerimoyle** [comment]
**1. Reset order specification**
Currently we have code in the validator which checks for resets with non-unique order values within a component.  It was my understanding that the order should be unique within the connected variable set, not the component?  

```
if (component->resetCount() > 0) {
        // Check for duplicate order values in resets
        std::vector<int> resetOrders;
        for (size_t i = 0; i < component->resetCount(); ++i) {
            ResetPtr reset = component->reset(i);
            int resetOrder = reset->order();
            if (reset->isOrderSet()) {
                if (std::find(resetOrders.begin(), resetOrders.end(), resetOrder) != resetOrders.end()) {
                    ErrorPtr err = std::make_shared<Error>();
                    err->setDescription("Component '" + component->name() + "' contains multiple resets with order '" + convertIntToString(resetOrder) + "'.");
                    err->setComponent(component);
                    err->setRule(SpecificationRule::RESET_ORDER);
                    mValidator->addError(err);
                } else {
                    resetOrders.push_back(resetOrder);
                }
            }
        }
        for (size_t i = 0; i < component->resetCount(); ++i) {
            ResetPtr reset = component->reset(i);
            validateReset(reset, component);
        }
    }
```

**2. test_variable component?**
In the old spec document it said that the `variable` needed to be present within the parent `component`. Is this also true for the `test_variable`, or can that be in another `component`?

**3. MathML stuff**
There is a function `validator->validateMath(const std::string &input, const ComponentPtr &component)` which tests the mathML against the `component` specified.  Do the contents of the `test_value` and `reset_value` blocks need to be specific to the parent component or can it refer to variables in other components?

**Comment by MichaelClerx** [comment]
For 2&3: Like all cellml constructs, it refers to variables in the local component only (although they may be defined in other components and then "connected" to the local versions)

What I _wanted_ to say in the earlier version, was that they had to be _defined_ (i.e. with mathml or an initial value) in the local component. But that's not allowed so we're not doing that!

**Comment by kerimoyle** [comment]
> that's not allowed

??

**Comment by MichaelClerx** [comment]
> > that's not allowed
> 
> ??

Eh, we've had discussions about this. Basically we're saying you can write algebraic ODEs etc. in CellML, so component A can say "X + Y  = 1 / Z", and then an entirely different model can have components defining 2 out of 3 of those, and that's all valid and something tools will have to put up with.

In practice most models and tools will prefer models written as assignments only, but for the standard it's easy enough to allow [DAEs](https://en.wikipedia.org/wiki/Differential-algebraic_system_of_equations), which have some important applications

**Comment by kerimoyle** [comment]
OK, that makes sense, thanks.  Any thoughts on the scope where `order`s must be unique?  Component-wide or equivalent-variable-set wide?

**Comment by MichaelClerx** [comment]
equivalent variable set. Are we unclear about this in the spec? We'll need to change that then...

**Comment by kerimoyle** [comment]
The spec seemed clear to me, but the existing tests reported something completely different (as in code above).  I think it would make more sense to move the testing for order uniqueness into the equivalent variable testing routines (like where units are tested etc) instead of just looking for different resets within a component with the same order...?

**Comment by MichaelClerx** [comment]
That sounds sensible to me. You need to have the DAG worked out before you can validate it

**Comment by kerimoyle** [comment]
DAG? Dependency Adjacency Graph? Or ...?

**Comment by MichaelClerx** [comment]
Directed acyclic graph, sorry! The variable equivalence graph is the name we were using I think

**Comment by kerimoyle** [comment]
@nickerso @agarny @hsorby Ready for review please.

File: src/printer.cpp
Comment (nickerso) [inline src/printer.cpp]: good to create an issue for this and list the issue here. Don't think this needs to hold up this PR.

File: src/printer.cpp
Comment (nickerso) [inline src/printer.cpp]: as above, good to link to the issue for future reference.

File: src/xmlattribute.cpp
Comment (nickerso) [inline src/xmlattribute.cpp]: seems like a random change that should be taken out of this PR.

File: src/xmlnode.cpp
Comment (nickerso) [inline src/xmlnode.cpp]: not clear why this is here? should remove from PR.

File: src/xmlnode.cpp
Comment (nickerso) [inline src/xmlnode.cpp]: and this one.

File: tests/parser/file_parser.cpp
Comment (nickerso) [inline tests/parser/file_parser.cpp]: Why are all these tests being commented out?

File: tests/parser/libxml_user.cpp
Comment (nickerso) [inline tests/parser/libxml_user.cpp]: more tests being commented out??

File: tests/parser/parser.cpp
Comment (nickerso) [inline tests/parser/parser.cpp]: Not sure if I have missed something in the discussion, but I'm expecting the test_value and rest_value to be an expression not an equation. i.e.,
```
<test_value>
   <math>
     <cn cellml:units="variable1units">3.4</cn>
   </math>
</test_value>
```

File: tests/printer/printer.cpp
Comment (nickerso) [inline tests/printer/printer.cpp]: not sure why this is here? another random whitespace addition to the PR that should be removed...

File: tests/reset/reset.cpp
Comment (nickerso) [inline tests/reset/reset.cpp]: better to remove old code than to comment out.

File: tests/reset/reset.cpp
Comment (nickerso) [inline tests/reset/reset.cpp]: unused?

File: tests/reset/reset.cpp
Comment (nickerso) [inline tests/reset/reset.cpp]: unused?

File: tests/reset/reset.cpp
Comment (nickerso) [inline tests/reset/reset.cpp:264]: unused?

File: tests/reset/reset.cpp
Comment (nickerso) [inline tests/reset/reset.cpp:134]: would be good to add tests with negative orders.

File: tests/validator/validator.cpp
Comment (nickerso) [inline tests/validator/validator.cpp]: would be good to test both options of either not setting the math or setting with an empty string.

**Review by nickerso — DISMISSED** [review]
Looking good. A bit more work to clarify things, test some more combinations, and put the parsing tests back in...

File: src/printer.cpp
Comment (kerimoyle) [reply src/printer.cpp]: Done, issue #364

File: src/printer.cpp
Comment (kerimoyle) [reply src/printer.cpp]: Done, issue #364

File: src/xmlattribute.cpp
Comment (kerimoyle) [reply src/xmlattribute.cpp]: Done

File: src/xmlnode.cpp
Comment (kerimoyle) [reply src/xmlnode.cpp]: Done (debugging leftovers!)

File: src/xmlnode.cpp
Comment (kerimoyle) [reply src/xmlnode.cpp]: Done (debugging leftovers)

File: tests/parser/parser.cpp
Comment (kerimoyle) [reply tests/parser/parser.cpp]: @nickerso I haven't seen anything either way, other than the contents of test_value and reset_value was to be mathml.  Is there a new version of the spec around resets yet? I've just been going from the use case document?  I'd imagine that the test_value needs to be an expression though? and that there could be cases where you'd want to reset to something related to another variable, so that the reset_value needs to be an expression too?  @MichaelClerx what are your thoughts?

File: tests/printer/printer.cpp
Comment (kerimoyle) [reply tests/printer/printer.cpp]: Done

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp]: Done

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp]: Done

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp]: Done

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp:264]: Umm .. no?

File: tests/parser/file_parser.cpp
Comment (kerimoyle) [reply tests/parser/file_parser.cpp]: Back in now

File: tests/parser/libxml_user.cpp
Comment (kerimoyle) [reply tests/parser/libxml_user.cpp]: Back in now

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp:134]: There will be more tests around orders to come when the order checking is in properly - at the moment in this PR its scope is not correct. Related to issue #326 (and discussion earlier in this PR).

File: tests/validator/validator.cpp
Comment (kerimoyle) [reply tests/validator/validator.cpp]: @nickerso  I've added a test for whitespace only, which returns an equivalent error to reset-value-not-set etc ... but am confused as to whether empty math blocks should be valid or not?  See issue #365 @MichaelClerx

**Comment by kerimoyle** [comment]
> Directed acyclic graph, sorry! The variable equivalence graph is the name we were using I think

@MichaelClerx Is it directed?  Ooops, I'd thought not ... ? Might need to adjust validation checks if so!

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp:134]: Order validation is now present in the other PR #350.  Tests will need to be tweaked when this is merged in, so if it's ok will add more tests when they don't need rewriting?

**Comment by kerimoyle** [comment]
Added validation for reset variable and test_variable being inside the same component, and the tests to cover them. Addresses issue #294.

File: tests/validator/validator.cpp
Comment (kerimoyle) [reply tests/validator/validator.cpp]: Done, including empty maths as per discussion in #365.

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp:134]: I've added a test with a negative order to the validator in this branch.

**Review by hsorby — CHANGES_REQUESTED** [review]
TEST(Validator, resets) is way too big and it needs to be broken down into smaller tests.

**Comment by kerimoyle** [comment]
> TEST(Validator, resets) is way too big and it needs to be broken down into smaller tests.

OK, will do.  Micro-tests coming up ... ;)

**Comment by kerimoyle** [comment]
@hsorby @agarny @nickerso I think I've addressed everything requested ... Good to go?

**Review by hsorby — CHANGES_REQUESTED** [review]
We also need to have some printer tests for making sure we print out resets as we expect.

**Comment by hsorby** [comment]
The broken down tests are much better for me I find them much easier to read and make sense of.

**Comment by kerimoyle** [comment]
Not at all DRY though :\

**Comment by hsorby** [comment]
Nope.  But it's testing so essentially you want to repeat yourself.

**Comment by kerimoyle** [comment]
> We also need to have some printer tests for making sure we print out resets as we expect.

We already test printing in the tests/reset.cpp for:

- printResetWithVariable
- printResetWithoutTestValue
- printResetWithOrder
- printResetWithNegativeOrder
- printResetWithOrderAndVariable
- addRemoveResetFromComponentMethods

... and in tests/printer.cpp for:
- printEmptyReset

I just adjusted the existing ones to the new format from what was there previously.  Are there tests that need to be added to these, or should the reset ones be in the printer.cpp file?

**Comment by hsorby** [comment]
There is no right place for them I was just trying to find them and thought they would be in the `tests/printer/printer.cpp` file but I didn't see any changes there.  I'll have another look.

File: tests/parser/parser.cpp
Comment (hsorby) [inline tests/parser/parser.cpp:1538]: Everything in this test should have an id, this doesn't.

File: tests/parser/parser.cpp
Comment (hsorby) [inline tests/parser/parser.cpp:1543]: This should also have an id.

File: tests/parser/parser.cpp
Comment (hsorby) [inline tests/parser/parser.cpp]: This shouldn't be here if it's just commented out.

File: tests/parser/parser.cpp
Comment (hsorby) [inline tests/parser/parser.cpp]: This has to go.

File: tests/parser/parser.cpp
Comment (hsorby) [inline tests/parser/parser.cpp]: Not a great job of removing???

**Review by hsorby — CHANGES_REQUESTED** [review]

**Comment by hsorby** [comment]
Trying to find resets API tests `appendResetValue` and so on but haven't managed it can someone point them out to me?

**Comment by hsorby** [comment]
Also do we want to have a `clearResetValue` API etc?

**Comment by kerimoyle** [comment]
> Also do we want to have a `clearResetValue` API etc?

I took the current implementation from what was there previously (see component's mathml functions?).   The same behaviour is found by calling `setResetValue("");` . Easy enough to add, but then will also be needed throughout as well?

**Comment by kerimoyle** [comment]
> Trying to find resets API tests `appendResetValue` and so on but haven't managed it can someone point them out to me?

TEST(Parser, repeatedMathParsePrintBehaviourWithReset)  calls code that uses them.  Does it need to be separately done?

File: tests/parser/parser.cpp
Comment (kerimoyle) [reply tests/parser/parser.cpp]: True dat ... I left it commented out because of the weirdness I was getting in parsing maths between two branches.

File: tests/parser/parser.cpp
Comment (kerimoyle) [reply tests/parser/parser.cpp:1538]: Is that a general thing for cellml stuff everywhere?  I thought ids were optional?

Comment (kerimoyle) [reply tests/parser/parser.cpp:1538]: Ignore me, just read the name of the test it's in ... :)

File: tests/parser/parser.cpp
Comment (kerimoyle) [reply tests/parser/parser.cpp]: Yep

**Comment by hsorby** [comment]
> > Also do we want to have a `clearResetValue` API etc?
> 
> I took the current implementation from what was there previously (see component's mathml functions?). The same behaviour is found by calling `setResetValue("");` . Easy enough to add, but then will also be needed throughout as well?

No I'm pretty sure we decided that setting empty is clear.  That is what everything else does.  Just reminding myself of that.

**Comment by hsorby** [comment]
> TEST(Parser, repeatedMathParsePrintBehaviourWithReset) calls code that uses them. Does it need to be separately done?

I would have that done they are unit tests each part should be tested on it's own.

File: tests/parser/parser.cpp
Comment (kerimoyle) [reply tests/parser/parser.cpp]: Deleted now that I know that the mathml printing/parsing is inconsistent for invalid maths.

File: tests/parser/parser.cpp
Comment (kerimoyle) [reply tests/parser/parser.cpp:1543]: Done

File: tests/parser/parser.cpp
Comment (kerimoyle) [reply tests/parser/parser.cpp:1538]: Done

**Comment by kerimoyle** [comment]
> I would have that done they are unit tests each part should be tested on it's own.

@hsorby Have added four new tests into tests/reset.cpp which deal with each bit individually: 
- testValueSetClear
- testValueAppend
- resetValueSetClear
- resetValueAppend

What's next ;) ?

File: tests/reset/reset.cpp
Comment (hsorby) [inline tests/reset/reset.cpp]: This should be moved into the test_utils.h file and then used throughout the tests.

File: tests/reset/reset.cpp
Comment (hsorby) [inline tests/reset/reset.cpp]: Not used so delete

File: tests/reset/reset.cpp
Comment (hsorby) [inline tests/reset/reset.cpp]: Not used so delete.

File: tests/reset/reset.cpp
Comment (hsorby) [inline tests/reset/reset.cpp:545]: Could also choose to append two empty maths blocks.

File: tests/reset/reset.cpp
Comment (hsorby) [inline tests/reset/reset.cpp]: Not used so delete

**Review by hsorby — CHANGES_REQUESTED** [review]
A few things to tidy up.

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp:545]: Oh, I could have chosen many, many other things ... :)

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp]: Good plan.

**Comment by kerimoyle** [comment]
I'm missing loads of variables which are declared but unused because there used to be a warning for this which I relied on ... why was it removed?  Seems a more useful warning than the your-test-failed-because-you-missed-a-single-space-tests. Or does anyone know how I do that locally?

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp]: Done

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp]: Done

File: tests/reset/reset.cpp
Comment (kerimoyle) [reply tests/reset/reset.cpp]: Done

**Comment by kerimoyle** [comment]
Good to go? @hsorby @nickerso @agarny

File: src/api/libcellml/reset.h
Comment (agarny) [inline src/api/libcellml/reset.h]: Use the infinitive, i.e. "Set" rather than "Sets". (Check the rest of the PR for similar occurrences.)

Comment (kerimoyle) [reply src/api/libcellml/reset.h]: Done :)

File: src/api/libcellml/reset.h
Comment (agarny) [inline src/api/libcellml/reset.h]: Replace with something like "@brief Append @p math to the existing testValue string for this @c Reset." (i.e. try to be consistent with what has been done for the `Component` class.) The same for the other API documentation.

Comment (kerimoyle) [reply src/api/libcellml/reset.h]: I think we should do this en masse for all of the documentation everywhere once all the pull requests have been merged.  There are lots of inconsistencies everywhere ... !

Comment (agarny) [reply src/api/libcellml/reset.h]: Agreed, ignore my comment for now.

**Comment by kerimoyle** [comment]
Nudge ... @hsorby @nickerso

**Review by hsorby — COMMENTED** [review]
Looks good to me I have made some documentation changes and made a pull request [here](https://github.com/kerimoyle/libcellml-keri/pull/19) to (maybe) address issues I have with the code.

**Comment by hsorby** [comment]
Sorry it took so long to get here.

**Comment by kerimoyle** [comment]
I've merged in @hsorby 's  changes now - all around comments.  Once all the outstanding PRs are into development it would be good to go through the comments again (see @agarny 's earlier comment here ...)

**Comment by kerimoyle** [comment]
So this one is good to merge now :) @hsorby @nickerso @agarny 
ps - not sure why the earlier reviews are still showing as requested, all items are finished.

**Review by hsorby — DISMISSED** [review]

**Comment by kerimoyle** [comment]
Help ... not sure what's going on here ... I've merged the updated development branch into this one just fine by following the instructions here, but now I get a unit test failure on the mergability test?  All tests are passing on my machine and my local branches merged without a hitch.  What does that mean? @hsorby @agarny

**Comment by agarny** [comment]
You might want to try to merge one more time. From experience, GitHug can sometimes report conflicts when there are actually none. It happened to me earlier today with [@hsorby's PR](https://github.com/agarny/libcellml/pull/5).

**Comment by kerimoyle** [comment]
Whew.  Should be good to merge now? @agarny @hsorby @nickerso

**Comment by kerimoyle** [comment]
Nudge nudge ... @hsorby @agarny @nickerso

**Comment by hsorby** [comment]
Looking at it now. :)

**Comment by hsorby** [comment]
There are some TODO's in the parser that mention removing becasuse they are validation.  I think they should definitely be removed and other cases also considered for removal.

**Comment by kerimoyle** [comment]
Thanks @hsorby I'll have another look.  There was a PR at some stage which removed these but I can't find it anymore ... :|

**Comment by kerimoyle** [comment]
... and which other cases are you talking about?

**Comment by hsorby** [comment]
Well ... there are some errors produced from invalid numbers for instance.  I can't decide if this is a more of a validation problem or something we should report that we have changed that may be a surprise to the user thus worthy of informing.

**Comment by kerimoyle** [comment]
I'm just going through the current `parser.cpp `file now and will remove anything that's not related to parsing.  In designing the Issue class I've split up the ideas based on whether people would get an informative error message from the validator alone - there were places in the parser where it made more sense to flag errors as they don't get propagated in a helpful way later.  I'll comment out all the tests and lines which are currently throwing errors here just to check that we're on the same page, then I'll remove them properly.  But .. this should all be part of the bigger discussion about what kind of things and where we want the `Issue` class to report ...?

**Comment by kerimoyle** [comment]
So I've gone back and checked the other messages, in particular around the number formatting stuff.  Even though it's technically validation to check the number format (see below), since the numbers are only _saved_ if they're valid, the validator has no way to access whatever the wrong value was since it wouldn't be saved.  This is one of those examples where once we have the `Issue` class it will make more sense - we can raise a warning in the parser instead of ignoring it altogether ... 

```
        if (attribute->isType("units")) {
            reference = attribute->value();
        } else if (attribute->isType("prefix")) {
            prefix = attribute->value();
        } else if (attribute->isType("exponent")) {
            if (isCellMLReal(attribute->value())) {
                exponent = convertToDouble(attribute->value());
            } else {
                ErrorPtr err = std::make_shared<Error>();
                err->setDescription("Unit referencing '" + node->attribute("units") + "' in units '" + units->name() + "' has an exponent with the value '" + attribute->value() + "' that is not a representation of a CellML real valued number.");
                err->setUnits(units);
                err->setRule(SpecificationRule::UNIT_EXPONENT);
                mParser->addError(err);
            }
        } else if (attribute->isType("multiplier")) {
            if (isCellMLReal(attribute->value())) {
                multiplier = convertToDouble(attribute->value());
            } else {
                ErrorPtr err = std::make_shared<Error>();
                err->setDescription("Unit referencing '" + node->attribute("units") + "' in units '" + units->name() + "' has a multiplier with the value '" + attribute->value() + "' that is not a representation of a CellML real valued number.");
                err->setUnits(units);
                err->setRule(SpecificationRule::UNIT_MULTIPLIER);
                mParser->addError(err);
            }
```

**Comment by kerimoyle** [comment]
So I've come full circle!  There are a lot of things throughout the parser which we need to figure out how to handle, and this can only be done in a helpful way once the `Issue` class is done.  I think they should stay as-is for the minute and be addressed separately so that it's not confused with the reset format which is the point of this PR?

**Comment by kerimoyle** [comment]
I think it's ready now.  Here's the long story:
- I've put back the previously commented-out error creation in the parser which deal with encapsulations as this PR should only deal with the resets.  The others will be dealt with in #314 and #368 
- I've removed the check for the order in a parsed reset being set, as this is purely validation and can be picked up later
- I've reinstated the checks for validation errors when the **real** cause of that error cannot be tracked downstream, ie: when an invalid piece of text is not saved, so the error that the validator can report is a lot less informative than that which the parser can report.  Things in this category are the number errors (the validator can only say that they're missing, not what's wrong with them, as their value is not saved).
- the other reset errors in the parser follow the same pattern as the parser errors elsewhere, most closely the parsing of encapsulations.  If we want to remove them from parsing the resets, we should consider removing them from parsing the encapsulations ... but that's a story for another PR like the one above.

Short story ... I think it's good to go now?  @hsorby @agarny @nickerso

**Comment by kerimoyle** [comment]
Nudge, nudge ... @hsorby @agarny @nickerso I think this one should go in before the #360 code generation one so that the generated code can be the right format?

**Comment by kerimoyle** [comment]
Pretty pretty please can someone review this ... @hsorby @agarny @nickerso

**Pull request closed without merging**
